lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -332,6 +359,19 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
       stringLeaf.append(iter->getSecond());
     }
 
+    assert(!(refSendPath.empty() && stringLeaf.empty()) &&
+           "nothing to index through");
+    for (auto subOp : llvm::reverse(indexing)) {
```

Comment:
Thanks for bringing this up, I agree this is a bit uncertain / not sufficiently pinned in, e.g., ABI specification, and is a fundamentally unsafe assumption presently.  Glad to have this mentioned publicly, thank you.  Yep we need to figure out how to make this reliable across compilation units, and likely won't be "one compiler appends the indexing it expects the other compiler lowered/emitted this as" but rather "ensure whatever's needed to use exported references is encoded for use across the boundary".

Even without ABI/cross-boundary concerns, this assumption is still problematic/employed even within the same design in order to rwprobe an element of an aggregate (that's not lowered), re:"hoping" at this point we can reliably produce strings that will work and be sure to target what we expect they do (re:reasoning through backend + what ExportVerilog produces -- simple example where this may fall down is if a field or something has to be renamed as part of legalization).

I don't like this assumption at all, to be honest, it's fragile and asking for problems.

However (and FWIW), we already make this same assumption elsewhere and... we need to fix all of this, particularly inner symbols through HW but even in FIRRTL (using inner symbols to handle this appears to be most likely path forward for this but not the only one).  I'm expecting/planning to address this "soon", working out how much infra cleanup/improvements to route through (as OM/Properties efforts need robust solutions to this same problem).

However inner symbols don't save us from needing to sort out the ABI details, it's not unsolvable just not pinned down sufficiently just yet.  Glad we're chatting about it!